### PR TITLE
specifying --force,podman exit with 0

### PR DESCRIPTION
**Pull request by aonoa**
Fixes: https://github.com/containers/podman/issues/14612

Signed-off-by: wufan [1991849113@qq.com](mailto:1991849113@qq.com)

**Does this PR introduce a user-facing change?**
The resource disappears at the end of the rm command, with --force podman exit with 0

### CODE REVIEW

**Comment by rhatdan** [comment]
First you have to sign your commits

git commit -a --amend -s
git push --force

File: cmd/podman/networks/rm.go [hunk 1]
```diff
@@ -71,6 +71,9 @@ func networkRm(cmd *cobra.Command, args []string) error {
 			fmt.Println(r.Name)
 		} else {
 			setExitCode(r.Err)
+			if networkRmOptions.Force && registry.GetExitCode() == 1 {
```
Comment (rhatdan) [inline cmd/podman/networks/rm.go:74]: This should only happen if the error is network does not exist, and this is the only error returned.

File: cmd/podman/pods/rm.go [hunk 2]
```diff
@@ -94,6 +94,9 @@ func removePods(namesOrIDs []string, rmOptions entities.PodRmOptions, printIDs b
 	responses, err := registry.ContainerEngine().PodRm(context.Background(), namesOrIDs, rmOptions)
 	if err != nil {
 		setExitCode(err)
+		if rmOptions.Force && registry.GetExitCode() == 1 {
+			registry.SetExitCode(define.ExecErrorCodeIgnore)
```
Comment (rhatdan) [inline cmd/podman/pods/rm.go:98]: This should only happen if the error is pod does not exist, and this is the only error returned.

File: cmd/podman/volumes/rm.go [hunk 3]
```diff
@@ -73,6 +73,9 @@ func rm(cmd *cobra.Command, args []string) error {
 			fmt.Println(r.Id)
 		} else {
 			setExitCode(r.Err)
+			if rmOptions.Force && registry.GetExitCode() == 1 {
```
Comment (rhatdan) [inline cmd/podman/volumes/rm.go:76]: This should only happen if the error is volume does not exist, and this is the only error returned.

File: cmd/podman/images/rm.go [hunk 4]
```diff
@@ -82,6 +82,9 @@ func rm(cmd *cobra.Command, args []string) error {
 			}
 		}
 		registry.SetExitCode(report.ExitCode)
+		if imageOpts.Force && registry.GetExitCode() == 1 {
```
Comment (rhatdan) [inline cmd/podman/images/rm.go:85]: This should only happen if the error is image does not exist, and this is the only error returned.

File: cmd/podman/containers/rm.go [hunk 5]
```diff
@@ -140,6 +140,9 @@ func removeContainers(namesOrIDs []string, rmOptions entities.RmOptions, setExit
 			}
 			if setExit {
 				setExitCode(r.Err)
+				if rmOptions.Force && registry.GetExitCode() == 1 {
```
Comment (rhatdan) [inline cmd/podman/containers/rm.go:143]: This should only happen if the error is container does not exist, and this is the only error returned.

**Comment by rhatdan** [comment]
You need to add tests for this on each one of the --rm types.

**Review by rhatdan — APPROVED** [review]

File: cmd/podman/containers/rm.go (hunk 5 above)
Comment (rhatdan) [reply cmd/podman/containers/rm.go:143]: The problem with that is the NOEXIST errors are all different.

Comment (Luap99) [reply cmd/podman/containers/rm.go:143]: How are they all different? The setExitCode() already take care of setting this, so all what needs to be done is check if --force is set

File: cmd/podman/root.go [hunk 6]
```diff
@@ -106,6 +106,9 @@ func Execute() {
 				fmt.Fprintln(os.Stderr, "Cannot connect to Podman. Please verify your connection to the Linux system using `podman system connection list`, or try `podman machine init` and `podman machine start` to manage a new Linux VM")
 			}
 		}
+		if registry.GetExitCode() == define.ExecErrorCodeIgnore {
+			registry.SetExitCode(0)
+		}
```
[the remark this answers is not in the document]
Comment (aonoa) [reply cmd/podman/root.go:111]: Yes, I also think definition 128 is not good, but any error will make the return value not 0
https://github.com/containers/podman/blob/3426d56b92be2ac1c3cc62fc578e9cb6d64aca81/cmd/podman/root.go#L100

**Comment by aonoa** [comment]
> 



> 请在提交签名时使用您的真实姓名，请参阅https://github.com/containers/podman/blob/main/CONTRIBUTING.md#sign-your-prs
> 
My user name is not my real name

**Comment by mheon** [comment]
I'll pass the DCO check manually, it's oversensitive about this.

**Comment by mheon** [comment]
You have a lint error

File: cmd/podman/root.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Luap99) [reply cmd/podman/root.go:111]: you could make the setExitCode function return the error and then set it to nil if it should not error

Or maybe more easy make --force imply --ignore, that should work al tleast for the commands who have --ignore.

**Comment by aonoa** [comment]
> @aonoa你能修复 lint 错误吗？
> 
> > 为隧道运行 golangci-lint
> > 构建标签隧道：apparmor，seccomp，selinux，linter，remote
> > 跳过的目录隧道：pkg/api，pkg/machine/e2e 
> > cmd/podman/images/rm.go:4：文件未`goimports`-ed（ goimports) 
> > "fmt"

I will add a blank line after the 4 line
https://github.com/golangci/golangci-lint/issues/138#issuecomment-667680383

**Comment by vrothberg** [comment]
Something wrong must have happened while rebasing.  Can you rebase another time and squash the commits into one?

**Comment by rhatdan** [comment]
@aonoa Can you rebase and clean this up?

**Comment by rhatdan** [comment]
I opened up a replacement for this PR. https://github.com/containers/podman/pull/14959 Should I close this PR?

**Pull request closed without merging**